src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharactersBitmap.cs

```diff
@@ -70,32 +70,21 @@ public AllowedCharactersBitmap Clone()
             return new AllowedCharactersBitmap((uint[])_allowedCharacters.Clone());
         }
 
-        // Determines whether the given character can be returned unencoded.
-        public bool IsCharacterAllowed(char character)
-        {
-            int codePoint = character;
-            int index = codePoint >> 5;
-            int offset = codePoint & 0x1F;
-            return ((_allowedCharacters[index] >> offset) & 0x1U) != 0;
-        }
-
         // Determines whether the given character can be returned unencoded.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public bool IsUnicodeScalarAllowed(int unicodeScalar)
-        {
-            int index = unicodeScalar >> 5;
-            int offset = unicodeScalar & 0x1F;
-            return ((_allowedCharacters[index] >> offset) & 0x1U) != 0;
-        }
+        public bool IsCharacterAllowed(char character) => IsUnicodeScalarAllowed(character);
 
+        // Determines whether the given character can be returned unencoded.
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public unsafe int FindFirstCharacterToEncode(char* text, int textLength)
+        public bool IsUnicodeScalarAllowed(uint unicodeScalar)
         {
-            for (int i = 0; i < textLength; i++)
+            uint index = unicodeScalar >> 5;
+            if (index >= (uint)_allowedCharacters.Length)
             {
-                if (!IsCharacterAllowed(text[i])) { return i; }
+                return false; // would be out-of-bounds read (unicodeScalar isn't BMP char)
             }
-            return -1;
+
+            return ((_allowedCharacters[index] >> ((int)unicodeScalar & 0x1F) /* offset */) & 1U) != 0;
```

Comment:
presumably _allowedCharacters only stores info about `2^21` characters but unicodeScalar is 32 bit, is there any chance this can throw IndexOutOfRange?

Reply:
`_allowedChars` is 8 KB (a bitmap holding 65,536 bits). We now perform a length check at the beginning of this method before accessing the array, so it should never go out of bounds.

The original implementation never would have gone out of bounds anyway (since the caller doesn't pass values outside the valid range), but adding the check is good practice nonetheless and it results in slightly more optimized codegen.